dot/state/service_test.go

```diff
@@ -141,11 +156,76 @@ func TestService_BlockTree(t *testing.T) {
 	require.Equal(t, stateA.Block.BestBlockHash(), stateB.Block.BestBlockHash())
 }
 
+func TestService_StorageTriePruning(t *testing.T) {
+	testDir := utils.NewTestDir(t)
+	defer utils.RemoveTestDir(t)
+
+	retainBlocks := 2
+	config := Config{
+		Path:     testDir,
+		LogLevel: log.LvlInfo,
+		Pruning: struct {
+			Mode              pruner.Mode
+			NumRetainedBlocks int64
+		}{
+			Mode:              pruner.Full,
+			NumRetainedBlocks: int64(retainBlocks),
+		},
+	}
+	serv := NewService(config)
+	serv.UseMemDB()
+
+	genData, genTrie, genesisHeader := newTestGenesisWithTrieAndHeader(t)
+
+	err := serv.Initialise(genData, genesisHeader, genTrie)
+	require.NoError(t, err)
+
+	err = serv.Start()
+	require.NoError(t, err)
+
+	var blocks []*types.Block
+	parentHash := serv.Block.GenesisHash()
+
+	totalBlock := 10
+	for i := 1; i < totalBlock; i++ {
+		block, trieState := generateBlockWithRandomTrie(t, serv, &parentHash, int64(i))
+
+		err = serv.Storage.blockState.AddBlock(block)
+		require.NoError(t, err)
+
+		old := trieState.Snapshot()
+
+		oldTs, err := rtstorage.NewTrieState(old)
+		require.NoError(t, err)
+
+		err = serv.Storage.StoreTrie(oldTs, block.Header)
+		require.NoError(t, err)
+
+		blocks = append(blocks, block)
+		parentHash = block.Header.Hash()
+	}
+
+	time.Sleep(2 * time.Second)
+
+	for _, b := range blocks {
+		_, err := serv.Storage.LoadFromDB(b.Header.StateRoot)
+		if b.Header.Number.Int64() >= int64(totalBlock-retainBlocks-1) {
+			require.NoError(t, err, fmt.Sprintf("Got error for block %d", b.Header.Number.Int64()))
+			continue
+		}
+		require.Error(t, err, fmt.Sprintf("Expected error for block %d", b.Header.Number.Int64()))
```

Comment:
can you explicitly check the error type? not a huge deal if not

Reply:
Done